test-data/unit/check-narrowing.test

```diff
@@ -1261,3 +1261,30 @@ def g() -> None:
         def foo(): ...
     foo()
 [builtins fixtures/dict.pyi]
+
+
+[case testNarrowingOptionalEqualsNone]
+from typing import Optional
+
+class A: ...
```

Comment:
What about the difference between classes with and without custom `__eq__`?

In the original issue `str` intances can never be equal to `None`. So, the original type narrowing was correct.

But, if some type has custom `__eq__` - we cannot be sure about this anymore.

Reply:
This comment is quite open-ended, there are so many nuances here that I'm left guessing what I should do.

This particular bug or fix isn't currently affected by `__eq__` method.

Are you asking that I should add tests with `str` also? And add `__eq__` method to `A`?

> In the original issue str intances can never be equal to None.

One could possibly have a subclass instance of `str` that overrides `__eq__`. But is that considered unsound usage? I dunno.

---

There is _another_ slightly related unsoundness case, when objects consider themselves equal to `None`. But that would be a separate bug and a harder discussion whether a fix would reduce the usefulness of narrowing.

https://gist.github.com/mypy-play/d6199eb69f69c933ced7da5c51c25a50

Reply:
This comment is good enough for me, thanks! :)

Reply:
Opened another issue for the described case: https://github.com/python/mypy/issues/15764